{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e0cca29",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "145fb952",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_LEfSe(abundance_df, group_series, accession):\n",
    "    \"\"\"\n",
    "    Generate input data in LEfSe format and save it as a .csv file.\n",
    "\n",
    "    args:\n",
    "        abundance_df (pd.DataFrame): Microbial abundance data (operational taxonomic units, listed as samples)\n",
    "        group_series (pd.Series): Sample grouping information (index is the sample name, value is the group)\n",
    "        accession (str): Project identifier, used for output naming\n",
    "        work_dir (str): Output directory\n",
    "    \"\"\"\n",
    "\n",
    "    # Merge all rows\n",
    "    group_series_df = pd.DataFrame(group_series).T\n",
    "    group_series_df.index = ['Group']\n",
    "    df_microbe = abundance_df.T[group_series_df.columns]\n",
    "    df_out = pd.concat([group_series_df, df_microbe], axis=0)\n",
    "    df_out.index.name = 'Subject'\n",
    "\n",
    "    return df_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d9f4901",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hock(\n",
    "    accession,\n",
    "    taxonomy,\n",
    "    work_dir,\n",
    "    meta_file,\n",
    "    host,\n",
    "    group_info,\n",
    "    group_filter=None,\n",
    "    output_path=None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Calculate the correlation of microbial and host gene expression (after LEfSe selection), and create a significance mask for gene/species selection.\n",
    "\n",
    "    returns:\n",
    "        DataFrame: id, species, gene, correlation, accession\n",
    "    \"\"\"\n",
    "\n",
    "    # 1. Obtain sample grouping\n",
    "    pie_meta = pd.read_csv(meta_file, dtype=str)\n",
    "    sub = pie_meta[(pie_meta['accession'] == accession)\n",
    "                   & (pie_meta['info'] == group_info)]\n",
    "    if group_filter:\n",
    "        for key, val in group_filter.items():\n",
    "            sub = sub[sub[key].isin(val)]\n",
    "\n",
    "    if sub.empty:\n",
    "        print(f\"⚠️ {accession}-{group_info} No qualifying samples.\")\n",
    "        return None\n",
    "\n",
    "    sample_list = sub['sample'].tolist()\n",
    "    sample_to_meta = dict(zip(sub['sample'], sub['value']))\n",
    "\n",
    "    # 2. Read the microbial abundance table.\n",
    "    summary_path = os.path.join(work_dir, host, f\"{accession}_summary.xlsx\")\n",
    "    if not os.path.exists(summary_path):\n",
    "        print(f\"❌ The file does not exist {summary_path}\")\n",
    "        return None\n",
    "\n",
    "    abundance = pd.read_excel(summary_path, sheet_name=taxonomy, index_col=0).T\n",
    "    abundance = abundance.loc[abundance.index.intersection(sample_list)]\n",
    "    if abundance.shape[0] < 2:\n",
    "        print(f\"⚠️ Insufficient sample size, skipping {accession}\")\n",
    "        return None\n",
    "\n",
    "    group_series = pd.Series(sample_to_meta)\n",
    "\n",
    "    preprocess_data = preprocess_LEfSe(abundance, group_series, accession)\n",
    "\n",
    "    preprocess_data.to_csv(os.path.join(output_path, f\"{accession}_lefse_input.tsv\"), sep='\\t')\n",
    "\n",
    "    return preprocess_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9300aec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = \"D:/Project/gutDBase/bracken_summary_GSE_filtered\"\n",
    "host = \"human\"\n",
    "meta_file = \"D:/Project/gutDBase/metadata/hum_pie.csv\"\n",
    "\n",
    "pie_meta = pd.read_csv(meta_file, dtype=str)\n",
    "accessions = pie_meta['accession'].unique()\n",
    "combo_table = pd.DataFrame({'accession': accessions, 'info': [\n",
    "                            'metaclass'] * len(accessions)})\n",
    "taxonomy = 'species'\n",
    "output_dir = os.path.join(\n",
    "    \"D:/Project/gutDBase/LEfSe_preprocess\", host)\n",
    "os.makedirs(output_dir, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c20c52ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, row in combo_table.iterrows():\n",
    "    acc = row['accession']\n",
    "    info = row['info']\n",
    "    print(f\"processing: {acc} - {info}\")\n",
    "    try:\n",
    "        hock(\n",
    "            accession=acc,\n",
    "            group_info=info,\n",
    "            taxonomy=taxonomy,\n",
    "            work_dir=work_dir,\n",
    "            meta_file=meta_file,\n",
    "            host=host,\n",
    "            output_path=output_dir)\n",
    "    except Exception as e:\n",
    "        print(f\"⚠️ exception occurred: {e}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dl",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
